### PR TITLE
cmake: Multi domain flashing using ninja flash or west flash

### PR DESCRIPTION
**Pull request by tejlmand**
This commit uses the zephyr_property_target:ZEPHYR_RUNNERS_YAML to
specify additional ZEPHYR_RUNNERS_YAML files to include when flashing
multi domain systems, such as the nRF53.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

-----------

Companion: https://github.com/nrfconnect/sdk-zephyr/pull/324

------------
This produces:
```
$ west flash --build-dir build/
-- west flash: rebuilding
[1/231] Preparing syscall dependency handling

[214/221] Linking C executable zephyr/zephyr_prebuilt.elf
Memory region         Used Size  Region Size  %age Used
           FLASH:      108112 B       256 KB     41.24%
            SRAM:       35633 B        64 KB     54.37%
        IDT_LIST:         136 B         2 KB      6.64%
[221/221] Generating zephyr/merged_nrf5340pdk_nrf5340_cpunet.hex
[17/231] Performing build step for 'spm_subimage'
[1/135] Preparing syscall dependency handling

[130/135] Linking C executable zephyr/zephyr_prebuilt.elf
Memory region         Used Size  Region Size  %age Used
           FLASH:         32 KB        32 KB    100.00%
            SRAM:        3856 B        64 KB      5.88%
        IDT_LIST:          40 B         2 KB      1.95%
[135/135] Linking C executable zephyr/zephyr.elf
[224/231] Linking C executable zephyr/zephyr_prebuilt.elf
Memory region         Used Size  Region Size  %age Used
           FLASH:      116228 B       984 KB     11.53%
            SRAM:       18526 B       384 KB      4.71%
        IDT_LIST:          72 B         2 KB      3.52%
[231/231] Generating zephyr/merged_domains.hex
-- west flash: using runner nrfjprog
-- runners.nrfjprog: Flashing file: /projects/github/ncs/nrf/samples/bluetooth/central_bas/build/zephyr/merged.hex
Parsing hex file.
Erasing page at address 0x0.
Erasing page at address 0x1000.
...
Erasing page at address 0x22000.
Erasing page at address 0x23000.
Erasing page at address 0x24000.
Applying system reset.
Checking that the area to write is not protected.
Programming device.
Applying pin reset.
-- runners.nrfjprog: Board with serial number 960133506 flashed successfully.
-- west flash: using runner nrfjprog
-- runners.nrfjprog: Flashing file: /projects/github/ncs/nrf/samples/bluetooth/central_bas/build/hci_rpmsg/zephyr/merged_nrf5340pdk_nrf5340_cpunet.hex
Parsing hex file.
Erasing page at address 0x1000000.
Erasing page at address 0x1000800.
Erasing page at address 0x1001000.
....
Erasing page at address 0x1019000.
Erasing page at address 0x1019800.
Erasing page at address 0x101A000.
Applying system reset.
Checking that the area to write is not protected.
Programming device.
Applying pin reset.
-- runners.nrfjprog: Board with serial number 960133506 flashed successfully.
```

### CODE REVIEW

File: cmake/multi_image.cmake [hunk 1]
```diff
@@ -27,6 +27,7 @@ if(IMAGE_NAME)
   share("set(${IMAGE_NAME}KERNEL_ELF_NAME ${KERNEL_ELF_NAME})")
   share("list(APPEND ${IMAGE_NAME}BUILD_BYPRODUCTS ${PROJECT_BINARY_DIR}/${KERNEL_HEX_NAME})")
   share("list(APPEND ${IMAGE_NAME}BUILD_BYPRODUCTS ${PROJECT_BINARY_DIR}/${KERNEL_ELF_NAME})")
+  share("set(${IMAGE_NAME}ZEPHYR_RUNNERS_YAML ${PROJECT_BINARY_DIR}/runners.yaml)")
```
Comment (oyvindronningstad) [inline cmake/multi_image.cmake:30]: Does this mean we don't need the modification of `yaml_contents` in partition_manager.cmake? https://github.com/nrfconnect/sdk-nrf/blob/master/cmake/partition_manager.cmake#L317

**Comment by oyvindronningstad** [comment]
I have one more small concern: This causes the chip to be reset twice, the first time in an "invalid" state. I think it's fine for the majority, but it might bite some users.

**Comment by tejlmand** [comment]
> I have one more small concern: This causes the chip to be reset twice, the first time in an "invalid" state. I think it's fine for the majority, but it might bite some users.

Thanks for sharing the consideration.
This behavior is similar to the previous `ninja flash` for multicore flashing using the depends in CMake.
But I think it is valuable if the reset between flash operation can be avoided.
Maybe by a setting in the runners.yaml, so that it is generic and the SoC that requires a reset between flashes can ensure that will happen.

File: cmake/multi_image.cmake (hunk 1 above)
Comment (tejlmand) [reply cmake/multi_image.cmake:30]: I believe that modification is due to the merging of mcuboot, spm, and app images.
Those are still flashed as one hex, so currently I believe the modification is still needed.

But if it is decided to flash all those images independently, then I believe this code could be removed, but I think care must be taken in that case regarding the signing and flashing of right signed image.

You and @hakonfam are probably the best people to consider this.

**Comment by tejlmand** [comment]
Closed by mistake: Reopened here: #2499

**Pull request closed without merging**